### ISSUE TITLE
REST api - clarifications needed for service entity example

### ISSUE THREAD

**Issue by abonas** [GoogleCloudPlatform/kubernetes#6439]
Using this example:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/examples/guestbook-go/v1beta3/guestbook-service.json

When created "as is", it works, however doing the same but omitting the 'kind' and 'apiversion'  causes it to fail with 422 with the following message:
""Service \"guestbook2\" is invalid: spec.sessionAffinity: required value""

1. If it already knows its a service (based on resource name and returned error msg), why there was a need to specify its kind in the first place from client side? Namespaces for example can be created np without those 2 properties.
2. why kind and api version removal causes an error regarding session affinity ? when they are in the input json, it is created successfully without specifying session affinity and its value is set to "None".

**Comment by yujuhong** [comment]
@abonas, after removing kind and apiversion, I got an error below that suggested kubectl wasn't able to tell what object this is:
`F0404 09:09:06.539786   30203 create.go:50] Client error: unable to recognize "examples/guestbook-go/v1beta3/guestbook-service.json": no object named "" is registered`

1. AFAIK, api version and kind *are* required to create an object. I tried creating a namespace without kind and apiversion and it returned the same "no object named" error. What command did you use to create the object?

2. Default values are applied to the object based on the apiversion it specifies. In your case, it's possible that the defaults are not being applied for some reason (e.g. lack of apivsersion?). I am not sure how you would get to this stage without encountering errors though.

**Comment by abonas** [comment]
> @abonas, after removing kind and apiversion, I got an error below that suggested kubectl wasn't able to tell what object this is:
> F0404 09:09:06.539786 30203 create.go:50] Client error: unable to recognize "examples/guestbook-go/v1beta3/guestbook-service.json": no object named "" is registered
> 
>     AFAIK, api version and kind are required to create an object. I tried creating a namespace without kind and apiversion and it returned the same "no object named" error. What command did you use to create the object?
> 
>     Default values are applied to the object based on the apiversion it specifies. In your case, it's possible that the defaults are not being applied for some reason (e.g. lack of apivsersion?). I am not sure how you would get to this stage without encountering errors though.

@yujuhong, I am not using kubectl, I am working with the REST api via http requests.
And as such, the url already contains the needed information:
If we examine this url for example:
http://localhost:8080/api/v1beta3/services, then it's clearly seen that both api version (v1beta3) and resource name (services) are specified, hence duplicating them in the request body seems redundant.
Moreover, the error message returned "Service \"guestbook2\" is invalid" implies that the server too is aware that this is an entity of type service.

**Comment by thockin** [comment]
It has always sort of very minorly annoyed me that some of the kubectl base
commands (get, watch) take a resource type on the cli and others (create,
update) dont.  It seems reasonable to me that the actual API server would
infer version and resource type from the URL, and that those fields might
be required by kubectl itself (or that we amend 'kubectl create' to take a
resource type).

**Comment by abonas** [comment]
> It has always sort of very minorly annoyed me that some of the kubectl base commands (get, watch) take a resource type on the cli and others (create, update) dont. It seems reasonable to me that the actual API server would infer version and resource type from the URL, and that those fields might be required by kubectl itself (or that we amend 'kubectl create' to take a resource type).

:+1:

**Comment by derekwaynecarr** [comment]
I like that I can let kubectl create multiple types of resources in one call.

**Comment by thockin** [comment]
Nothing proposed would stop that...

**Comment by bgrant0607** [comment]
@abonas You've previously reported this issue: #3000. We don't disagree. We just haven't got to it yet. PRs welcome.

Do you mind if we close this issue as a duplicate of #3000?

**Comment by abonas** [comment]
I think there were slightly different symptoms there as well as a
discussion re sending v1 object to v3 api. But if you view them as one
problem and one solution for both,  I have no objection closing one as dup.

**Comment by bgrant0607** [comment]
Ok, I'll leave this open re. the obscure error message.

**Comment by pythonpowerd** [comment]
+1 to fix this bug. I spend some time to figure out why my service is not created thru REST until i explicitly set apiVersion in request body or explicitly set spec.type: 'ClusterIP' which is default.
I think missing fields(like kind, apiVersion etc.) should be populated from url

**Comment by bgrant0607** [comment]
Applies to namespace, also.

**Comment by bgrant0607** [comment]
Related to this, we need a better way to communicate attributes to handlers in apiserver. Namespace is passed in context. The admission controller now gets a bundle of attributes. Query parameters are passed via specialized Go structs. Other metadata is assumed to be explicit in the resource body for most calls, but this is not always true, such as for subresources.

cc @lavalamp

**Comment by vefimova** [comment]
Hi, I'm working on this one

**Comment by vefimova** [comment]
@bgrant0607 AFAIU after https://github.com/kubernetes/kubernetes/pull/10200/commits/664d20c572498fd443a678160ff6a49fd9d4d22a was committed there is no issue anymore (obscure error message): as resources will be created successfully via HTTP request without Kind and API version set. Correct me if I'm wrong.

**Comment by caesarxuchao** [comment]
@vefimova, yes, the original issue is solved. I think @bgrant0607 keeps it open for "we need a better way to communicate attributes to handlers in apiserver."

**Comment by vefimova** [comment]
Hi, @caesarxuchao.
So regarding to “we need a better way to communicate attributes to handlers in apiserver” I want to clarify what exactly it assumes. From the above comment from @bgrant0607 I can see the following items:

1. Namespace is passed in context - it says about api.Context and its key/value interface? If yes,It’s the right place to have attributes, right?
2. the admission controller now gets a bundle of attributes - there is a attributesRecord struct which populated for further work, we want to have all needed values within api.Context and populate struct using it?
3. Query parameters are passed via specialized Go structs - have them to be passed in api.Context?
4. Other metadata is assumed to be explicit in the resource body for most calls, but this is not always true, such as for subresources -afaiu it says here about subresources which actually perform some actions, like exe, attach, port-forward, which uses POST http request with parameters set in url and with empty message body. So let’s say apiserver gets ‘exe’ for pod, then we’d like to have all resource body relevant to pod been retrieved and be available thru key/value interface of api.Contex?

**Comment by bgrant0607** [comment]
That error is emitted by kubectl rather than by apiserver, but I suppose this could be lumped into #5889
